feat(http): allow `HttpClient` to cache requests

alan-agius4 opened the pull request:
This commit adds a new option for `provideHttpClient` called `withTransferCache()`. When this option is passed, requests done on the server are cached and reused during the  bootstrapping of the application in the browser thus avoiding duplicate requests and reducing load time.

This is the same as `TransferHttpCacheModule` in https://github.com/angular/universal/blob/main/modules/common/src/transfer_http.ts and is part of the effort to move "common" logic that can be used without "@nguniversal/common" into the framework.

alan-agius4 commented on packages/common/http/src/provider.ts:
```diff
@@ -6,12 +6,13 @@
  */
 
-import {EnvironmentProviders, inject, InjectionToken, makeEnvironmentProviders, Provider} from '@angular/core';
+import {ApplicationRef, EnvironmentProviders, inject, InjectionToken, makeEnvironmentProviders, Provider, ɵTransferState as TransferState} from '@angular/core';
```
@AndrewKushnir, should we make `ɵTransferState` from `@angular/core` as public, re-export it from `@angular/platform-browser` and deprecate the latter export? 

Otherwise users might end up with 2 copies of `TransferState` which will increase the bundle size slightly.

PS: happy to do the change myself.

AndrewKushnir reviewed (approved):
@alan-agius4 looks great! 👍

atscott commented on packages/common/http/src/transfer_cache.ts:
```diff
@@ -0,0 +1,116 @@
+  // Stop using the cache if the application has stabilized, indicating initial rendering is
+  // complete.
+  let isCacheActive = true;
+  appRef.isStable.pipe(filter((isStable) => isStable), take(1))
+      .subscribe((stable) => {
+        isCacheActive = !stable;
+      })
+      .unsubscribe();
```
I'm a bit confused by this. This subscribes then unsubscribes immediately meaning that only the first synchronous value of `isStable` gets used (this example never logs anything: https://stackblitz.com/edit/rxjs-5xnvhe?devtoolsheight=60&file=index.ts)

The code for the `isStable` observable used in `ApplicationRef` is here:
https://github.com/angular/angular/blob/69a6e09e7eb13b592fb5430cf83800d9d73aea71/packages/core/src/zone/ng_zone.ts#L530-L574

I think what you effectively have is just the value of `isCurrentlyStable` (`zone.isStable && !zone.hasPendingMacrotasks && !zone.hasPendingMicrotasks`). I think it would make more sense to just use that line instead because it's more clear what value is being used.

atscott replied:
Ah, I missed the comment about this coming from https://github.com/angular/universal/blob/main/modules/common/src/transfer_http.ts. I think what you want here is 
```
 appRef.isStable.pipe(filter((isStable) => isStable), take(1))
      .subscribe(() => {
        isCacheActive = false;
      });
```

alan-agius4 replied:
That would work, my concern was that we'll leave open subscriptions until the app is first sthbilized.

atscott replied:
Ah, right because this isn't a class like before and gets called on each request. FWIW, I still think it would be better to update this to be `zone.isStable && !zone.hasPendingMacrotasks && !zone.hasPendingMicrotasks` to remove rxjs from the equation since you're really not trying to use the stream.

alan-agius4 replied:
Correct. 

I would love to get rid of RxJs, but `zone.isStable && !zone.hasPendingMacrotasks && !zone.hasPendingMicrotasks` results in a different behaviour. As this only considers the current state.  What I want is if it was ever stable, even though it is not right now, cache is not used.

atscott replied:
I'm not sure the current behavior does this either, actually. It's still only getting the current synchronous value of `appRef.isStable` so if it became stable and then unstable again before the interceptor runs, it will still not know that it was ever stable, right?

alan-agius4 replied:
I think you meant, it is app is unstable (cache is enabled), it becomes stable (cache should be disabled), but since we got the value synchronous before the interceptor runs the response would be returned still from cache when it should not. 

This is possible, but I think it's quite an edge case for this to happen that an interceptor is invoked and not run or I am missing something? 

If you have some ideas on how to solve this, happy to hear it :).

atscott replied:
I created an attempt to do this here: https://github.com/angular/angular/commit/5804eec1b41fad812df8d049fbf9c754eb23123e

You kind of need the subscription to live somewhere and stay subscribed to the stable observable so you don't miss anything from the stream that you care about.

alan-agius4 replied:
That’s brilliant, I love it 🤩, Will incorporate it tomorrow as it’s pretty late for me today.

Thank you @atscott.

bougwal commented on packages/common/http/src/transfer_cache.ts:
```diff
@@ -0,0 +1,116 @@
+import {HttpEvent, HttpResponse} from './response';
+
+interface TransferHttpResponse {
+  body?: any;
+  headers?: Record<string, string[]>;
+  status?: number;
+  statusText?: string;
+  url?: string;
+  responseType?: HttpRequest<unknown>['responseType'];
```
nit: since all keys are optional, maybe it would be more descriptive below to use TypeScript's utility types:  `Partial<TransferHttpResponse>` 

```suggestion
  body: any;
  headers: Record<string, string[]>;
  status: number;
  statusText: string;
  url: string;
  responseType: HttpRequest<unknown>['responseType'];
```

alan-agius4 commented on packages/common/http/src/transfer_cache.ts:
```diff
@@ -0,0 +1,116 @@
+  return headersMap;
+}
+
+export function makeCacheKey(request: HttpRequest<any>): StateKey<TransferHttpResponse> {
```
I agree, in this case we can use crypto API, since we can generate the hash in async manner.

alan-agius4 replied:
Switched to use the same hashing logic as the one used for component ids, as using web crypto API requires some workarounds for cross platform usage.

alan-agius4 commented on packages/common/http/src/transfer_cache.ts:
```diff
@@ -0,0 +1,116 @@
+  return next(req).pipe(
+      tap((event: HttpEvent<unknown>) => {
+        if (event instanceof HttpResponse) {
+          transferState.set<TransferHttpResponse>(storeKey, {
```
Updated the interface, as actually body an header cannot be optional.

bougwal replied:
@alan-agius4 I synced up Universal's interceptor's `TransferHttpResponse` to reflect this comment here. A tiny PR is raised under https://github.com/angular/universal/pull/3057

AndrewKushnir commented on packages/common/http/src/provider.ts:
```diff
@@ -233,3 +236,26 @@ export function withRequestsMadeViaParent(): HttpFeature<HttpFeatureKind.Request
+ *
+ * @see provideHttpClient
+ */
+export function withTransferCache(): HttpFeature<HttpFeatureKind.TransferCache> {
```
As discussed offline, this feature should be at the app root level (not a part of HttpClient).

AndrewKushnir commented on packages/common/http/src/transfer_cache.ts:
```diff
@@ -0,0 +1,142 @@
+  const appRef = inject(ApplicationRef);
+
+  let isCacheActive = true;
+  appRef.isStable
```
We can also take the `InitialRenderPendingTasks` status into account as well (similar to [utils.ts in the platform-server](https://github.com/angular/angular/blob/main/packages/platform-server/src/utils.ts#L60)).

AndrewKushnir replied:
It also looks like we'd need to create a helper method for this (as a followup), since we use this "signal" in multiple places now.

AndrewKushnir replied:
We can probably change `ENVIRONMENT_INITIALIZER` -> `APP_BOOTSTRAP_LISTENER`, which is used in the hydration code:
https://github.com/angular/angular/blob/main/packages/core/src/hydration/api.ts#L163

In this case, the `InitialRenderPendingTasks` state would reflect the state of an initial navigation. WDYT?

alan-agius4 replied:
Yeah that should work.

alan-agius4 replied:
@AndrewKushnir  PTAL

AndrewKushnir commented on packages/common/http/src/transfer_cache.ts:
```diff
@@ -0,0 +1,142 @@
+  appRef.isStable
+      .pipe(
+          filter((isStable) => isStable),
+          take(1),
```
```suggestion
          take(1)
```

AndrewKushnir commented on packages/common/http/src/transfer_cache.ts:
```diff
@@ -0,0 +1,142 @@
+    // Request found in cache. Respond using it.
+    let body: ArrayBuffer|Blob|string|undefined = response.body;
+
+    switch (response.responseType) {
+      case 'arraybuffer':
+        body = new TextEncoder().encode(response.body).buffer;
+        break;
+      case 'blob':
+        body = new Blob([response.body]);
+        break;
+    }
```
I'm wondering if there is an existing helper (or a logic that we can extract) somewhere in the http package to do this?

alan-agius4 replied:
There is a similar helper in https://github.com/angular/angular/blob/1dc8480bf4ac7bbd34f3413a6c5c7cafa30ee15f/packages/common/http/testing/src/request.ts#L200 but it is more involved.

AndrewKushnir reviewed (commented):
Looks great! 👍  Just left a few comments.

alan-agius4 committed:
feat(http): allow `HttpClient` to cache requests

This commit adds a new option for `provideHttpClient` called
`withHttpTransferCache()`. When this option is passed, requests done on the server are cached and reused during the  bootstrapping of the application in the browser thus avoiding duplicate requests and reducing load time.

This is the same as `TransferHttpCacheModule` in https://github.com/angular/universal/blob/main/modules/common/src/transfer_http.ts

alan-agius4 commented on packages/common/http/src/interceptor.ts:
```diff
@@ -145,12 +145,21 @@ function chainedInterceptorFn(
+/**
+ * A multi-provided token of `HttpInterceptorFn`s that are only set in root.
+ */
+export const HTTP_ROOT_INTERCEPTOR_FNS = new InjectionToken<HttpInterceptorFn[]>(
```
Is there a cleaner way?

atscott replied:
Instead of having a default factory, can you just inject it optionally below? 

`this.injector.get(HTTP_ROOT_INTERCEPTOR_FNS, [])`

alan-agius4 replied:
Good call.

alan-agius4 committed:
fixup! feat(http): allow `HttpClient` to cache requests

alan-agius4 committed:
fixup! feat(http): allow `HttpClient` to cache requests

alan-agius4 commented:
**Caretaker note:** G3 failure is pre-existing.

AndrewKushnir reviewed (approved):
Looks great, thanks Alan! 👍

dylhunn commented:
This PR was merged into the repository by commit aff15129501511569bbb4ff6dfcb16ad1c01890d.
